cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -122,27 +136,86 @@ public static void validateConfig(SecureStore secureStore, SourceControlConfig s
   }
 
   /**
-   * Initializes the Git repository by cloning remote.
+   * Returns the <a href="https://git-scm.com/docs/git-hash-object">Git Hash</a>
+   * of the requested file path in the provided commit. For symlinks, it returns the hash of
+   * the target file. This ensures that the file hash of a symlink is equal to the hash of the target file. It
+   * doesn't return the hash for directories. If the directory contains symlinks, we can't depend on the
+   * directory hash to ensure the contents are unchanged.
    *
+   * @param relativePath The path relative to the repository base path (returned by
+   *                     {@link RepositoryManager#getRepositoryRoot()}) on the filesystem.
+   * @param commitHash   The commit ID hash for which to get the file hash.
+   * @return The git file hash of the requested file.
+   * @throws IOException              when file or commit isn't found, there are circular symlinks or other file IO
+   *                                  errors.
+   * @throws IllegalArgumentException when the provided file isn't a regular file, eg: directory.
+   * @throws NotFoundException        when the file isn't committed to Git.
+   * @throws IllegalStateException    when {@link RepositoryManager#cloneRemote()} isn't called before this.
+   */
+  public String getFileHash(Path relativePath, String commitHash) throws IOException, NotFoundException,
+    GitAPIException {
+    validateInitialized();
+    // Save changes before checking out the requested commit.
+    ObjectId previousCommit = resolveHead();
+    // Stash changes in current working directory. Stash ID will be null if working directory is clean.
+    RevCommit stashID = git.stashCreate().setIncludeUntracked(true).call();
+
+    try {
+      ObjectId commitId = ObjectId.fromString(commitHash);
+      // Each commit points to a tree that contains the files/directories as nodes. The ObjectId (hashes) of the nodes
+      // represent the state of a file at that commit.
+      RevCommit commit = new RevWalk(git.getRepository()).parseCommit(commitId);
+      // Checkout the requested commit.
+      git.checkout().setName(commitHash).call();
+
+      // Git considers symlinks as separate files. So the hash of the symlink will be different from the target file.
+      // Resolve path in case of symlinks. This ensures the hash of the target file is returned.
+      Path realPath = getRepositoryRoot().resolve(relativePath).toRealPath();
+      // We should not be comparing hashes of directories. If the directory contains symlinks, we can't depend on the
+      // directory hash to ensure the contents are unchanged.
+      if (!Files.isRegularFile(realPath)) {
+        throw new IllegalArgumentException(String.format("Path %s doesn't refer to a regular file.",
+                                                         realPath.toAbsolutePath()));
+      }
+      Path realRelativePath = getRepositoryRoot().relativize(realPath);
```

Comment:
Since `getRepositoryRoot()` might also return a symlinked path, e,g on macOS the `/tmp` directory is auto linked to `/private`: https://apple.stackexchange.com/questions/1043/why-is-tmp-a-symlink-to-private-tmp

We might need to do `Path realRelativePath = getRepositoryRoot().toRealPath().relativize(realPath);`

@arjan-bal If you agree, let's make add the change in the pushApps PR and leave this PR as is

Reply:
Nice catch. Acknowledged, not making the change in this PR.